modal/_utils/async_utils.py

```diff
@@ -510,3 +510,37 @@ async def async_zip(*inputs: Union[AsyncIterable[T], Iterable[T]]) -> AsyncGener
             yield tuple(items)
         except StopAsyncIteration:
             break
+
+
+async def async_merge(*inputs: Union[AsyncIterable[T], Iterable[T]]) -> AsyncGenerator[T, None]:
+    queue: asyncio.Queue[Tuple[int, Union[T, Exception]]] = asyncio.Queue()
+
+    async def producer(index: int, iterable: Union[AsyncIterable[T], Iterable[T]]):
```

Comment:
nit: `index` could be `producer_id` and `active_inputs` could be `active_producers` to make the intention a bit clearer (index could be misunderstood to have to do with the index of an item in the input iterator etc.)